spec: sharded archiving draft (#192)
nazar-pc/abundance: specs/01-sharded-archiving.md

# Sharded Archiving

This is the subprotocol responsible for creating a global canonical history of the whole system, and
of creating the history records that will eventually be archived and become part of farmers' plots.

Shards in the multi-shard Subspace protocol operate independently as parallel chains: producing new
blocks, executing transaction in parallel and independently generating segments of their history
that will eventually become part of the global history stored in the beacon chain.

### Data Structures

This section collects a list of data structures required for the operation of the protocol:

> - `segment_commitments[]`: hash map of KZG commitments computed over pieces in a segment and stored

Author: Great! This is the legacy notation from before we had made the decision. Renamed accordingly.

  in the runtime of each shard. This data structure represents the segment history of the shard and

>   include both, local segment commitments, and segment commitments from child shards (through

Author: Good catch! I forgot to update this after the change to Merkle Trees instead of KZG, and the commitment of segments directly to the beacon chain (editted the description accordingly).

My idea here was to store this information however is currently done for `segment_commitments[]` now in Subspace. From skimming through the code and the spec my understanding is that this is stored on-chain as state in the state of the blockchain, but please correct me if this is not the case.

Let me know if I need to elaborate on this in the description after my edit.

  `SuperSegment`s). Each of them should be accordingly tagged as such.

> - `shard_blocks[]`: hash map of blocks committed from child shards. It stores the history of the

Author: Same answer as above, I was thinking this to be a convenient state object stored on-chain an updated when a block includes a `BlockInfo` or a  new segment, but happy to be stand corrected.

> - `shard_blocks[]`: hash map of blocks committed from child shards. It stores the history of the

Reviewer: Blocks are different, how do you plan to use this?

  shard in the parent, until these blocks are made available in a new segment.

> - `recorded_history_buffer`: Buffer storing segments (i.e. the local history) of a local shard

Author: I understand. I updated the text so that this is clear. The reason why I added this data structure is because it felt useful to understand the flow of the different data structures in the operation of the protocol. In any case, if you feel this is redundant (or unnecessary) happy to remove it.

### Preparing the history of a shard

#### 1. Genesis Block Archiving

- Input: Genesis block of the shard.
- The genesis block is archived as soon as it is produced. We extend the encoding of the genesis

>   block with extra pseudorandom data up to `RECORDED_HISTORY_SEGMENT_SIZE`, such that the very first

Author: Linked to that doc for additional context.

  archived segment can be produced right away, bootstrapping the farming process. This extra data is

>   added to the end of the SCALE-encoded block, so during decoding of the genesis block it'll be

Author: This makes sense. I updated it to mentioned encoded blocks but not specify what type of encoding.

TBH, I personally hate SCALE-encoding. I rather have something typed like CBOR even if it is a bit less efficient that having a context-aware encoding (which can make things a bit of a mess depending on the context).

>   added to the end of the SCALE-encoded block, so during decoding of the genesis block it'll be

Reviewer: I actually like SCALE quite a bit! It is **very** simple (I can encode/decode it with or without tooling just fine, and in fact archiver and data retrieval both rely on its rules to make sense of the contents), strictly typed, very efficient (basically as compact as it gets) and integrates with Rust beautifully through the derive macros. Encoding/decoding requires little code and little time, but it is still not zero copy in many cases.

- Each segment will have the parent `segment_header` included as the first item. Each
  `segment_header` includes the hash of the previous segment header and the segment commitment of
  the previous segment. Together, segment headers form a chain that is used for quick and efficient
  verification of pieces corresponding to the actual archival history of the blockchain.
- Segment items are combined into segments. Each segment contains at least two of the following
  `SegmentItem`s, in this order:
- The previous segment’s SegmentHeader
  - BlockContinuation, if remained from the previous segment
  - Block(s), as many as fit fully into the current segment, may be none
  - BlockStart, if the next block doesn't fit within the current segment
  - Padding (zero or more) in case padding bytes are necessary to complete the segment to
    `RECORDED_HISTORY_SEGMENT_SIZE`
- Output: Buffered SegmentItems.

#### 3. Block Info Construction

- Input: Blocks produced in the shard.
- When a new block is produced and included to the chain in a shard, farmers (and any other node in
  the system), extend the block using an error correction scheme to generate parity shares:

>   1. The block is split into `M = ceil(block_size/l)` shares of `l` bytes.

Reviewer: I would prefer to not use single-letter and otherwise cryptic variables/parameters/constants. For example this is much more readable to me:
```suggestion
  1. The block is split into `num_shares = ceil(block_size / SHARE_SIZE)` shares of `SHARE_SIZE` bytes.
```

`SHARE_SIZE` is uppercase because it is a constant, the rest are runtime variables.

It is a lot more verbose, but requires much less brain power to process and it is easier to search other uses of the same variable (like `num_shares`) in specs and code. Something like `l` will have huge number of matches in basically every file, making it very painful to find where it is used.

  2. Each share is arranged into a different column of a `M x l = ceil(block_size/l) x l` matrix.
  3. Shares are erasure coded column-wise with `extend(column, ERASURE_CODING_RATE)`. This
     effectively doubles the number of rows and, hence, the shares per block

>      to `total_shares = 2*M`.

Reviewer: This doesn't really compute in my head in a straightforward way.

When I think of a block sliced into shared that are then erasure coded, I imagine this:

| column_0 | column_1 |
|-|-|
| source_share_0 | source_share_1 |

Turned into this:

| column_0 | column_1 | column_2 | column_3 |
|-|-|-|-|
| source_share_0 | source_share_1 | parity_share_0 | parity_share_1 |

With `source_share0 || source_share1` being the block (+padding).

But you clearly meant something different since you're talking about doubling the number of rows.

  `segment_header` includes the hash of the previous segment header and the segment proof of the
  previous segment. Together, segment headers form a chain that is used for quick and efficient
  verification of pieces corresponding to the actual archival history of the blockchain.
- Segment items are combined into segments. Each segment contains at least two of the following
  `SegmentItem`s, in this order:
- The previous segment’s SegmentHeader
  - BlockContinuation, if remained from the previous segment
  - Block(s), as many as fit fully into the current segment, may be none
  - BlockStart, if the next block doesn't fit within the current segment
  - Padding (zero or more) in case padding bytes are necessary to complete the segment to
    `RECORDED_HISTORY_SEGMENT_SIZE`
- Output: Buffered SegmentItems.

#### 3. Block Info Construction

- Input: Blocks produced in the shard.
- When a new block is produced and included to the chain in a shard, farmers (and any other node in
  the system), extend the block using an error correction scheme to generate parity shares:

  1. The block is split into `source_shares = ceil(block_size/SHARE_SIZE)` shares of `SHARE_SIZE`
     bytes.

>   2. Each share is arranged into a different column of a

Author: Changed the description and the visualisations to align with how it seems more natural from an implementation perspective (the literature generally uses matrices to determine the space encoding, but I see your point, this may be easier for the spec).

     `source_shares x SHARE_SIZE = ceil(block_size/SHARE_SIZE) x SHARE_SIZE` matrix.

```
| source_share_0 |
| source_share_1 |
...
| source_share_M |
```

3. Shares are erasure coded column-wise with `extend(column, ERASURE_CODING_RATE)`. This effectively
   doubles the number of rows and, hence, the shares per block to `total_shares = 2*source_shares`.

> - The erasure coding rate is `ERASURE_CODING_RATE = 2`, which means that the number of shares

Author: Good catch, fixed.

>   generated is `total_shares = 2 * M`.

Author: Oh, man! Really good catch. Thanks!

```
| source_share_0 |
| source_share_1 |
...
| source_share_M |
| parity_share_0 |
| parity_share_1 |
...
| parity_share_M |
```

- The erasure coding scheme used is Reed-Solomon, which allows to recover the original data from any
  `M` shares.
- The size of each share is `SHARE_SIZE = block_size / source_share`.
- The number of chunks per share is `num_chunks_per_share = SHARE_SIZE / BLOCK_CHUNK_SIZE`, where
  `BLOCK_CHUNK_SIZE` is the size of each chunk.

4. Shares themselves are divided into chunks of size `BLOCK_CHUNK_SIZE`, in a way that `SHARE_SIZE`
   must be multiple of `BLOCK_CHUNK_SIZE`, `SHARE_SIZE = n x BLOCK_CHUNK_SIZE`.
5. For each row, a proof of the row chunks is created by building a Merkle Proof of the individual

>    chunks in the share, `Cbs_i`.

Author: fixed.

- Output: Error corrected encoded blocks, and `BlockInfo` to be submitted to the parent shard.

#### 4. Block Info Submission

- Input: `BlockInfo` from a child shard.
- Any node in the shard (from farmers to full-nodes) are allowed to create the `BlockInfo`
  transaction that submits the block information into the parent's shard history.
- The corresponding piece of the block should be stored in the shard and made retrievable via a
  peer-to-peer network protocol (DSN).
- When the transaction in the parent
- Output: Transaction in the parent shard's mempool.

#### 5. Shard block proof in parent

- Input: Transaction with `BlockInfo` from a child shard.
- The execution of the `BlockInfo` transaction when included in a block of the parent shards commits
  the block from the child shard to the history of the parent by including it on-chain into the
  `shard_blocks[]` table.
- To verify the `BlockInfo`, nodes in the parent shard:
  - Ensure `block_index` is the immediate block after last committed block number in the parent
    shard.
- Output: `BlockInfo` from child shard included in `shard_blocks[]` table of the parent shard.

> #### 6. Segment Construction

Reviewer: Left two comments, but this whole section needs to be updated

- Input: Buffered data of size `RAW_RECORD_SIZE`.
- When the buffer (after encoding) contains enough data to fill a record of `RAW_RECORD_SIZE` bytes,
  it is archived:

>   1. Split the record into `d = 2^15 =NUM_CHUNKS` chunks of size `SAFE_BYTES`.

Reviewer: Sorry, but `SAFE_BYTES` is gone as of https://github.com/nazar-pc/abundance/pull/182

Also:
```suggestion
 1. Split the record into `NUM_CHUNKS = 2^15` chunks 32 bytes each.
```

  2. Perform the same operations to extend with parity chunks and commit to the chunks as was done
     for the block. This is done by creating a Merkle tree of the chunks of a record, and committing
     to it.
- Erasure code records column-wise with `extend(column, ERASURE_CODING_RATE)`. This effectively
  doubles the number of rows and thus, records per segment to `NUM_PIECES` (256).
- For each row, a proof of the row chunks is created by building a Merkle Proof of the individual
  chunks in the share, `Cs_i`, along with their proof `ws_i`.

> - For each record, form a `piece = record || record_proof || record_proof`, `(ri∣∣Ci∣∣πi)`

Reviewer: This says two record proofs, but it is actually different:
https://github.com/nazar-pc/abundance/blob/2760b0ad8f67ae92dc33a22f5f68f161371a80db/subspace/crates/subspace-core-primitives/src/pieces.rs#L1023-L1045

Author: This is typo. Thanks for flagging and sharing the reference code.

- The segment now consists of `NUM_PIECES` records of roughly 1MiB each
  (`32 bytes * 2^15 chunks + 32 bytes proof + ~120 bytes proof`), `NUM_PIECES` piece
  proofs, `NUM_PIECES` proofs of 32 bytes each and one 32-byte `segment_root`.
- New pieces need to be stored by farmers and made available for retrieval by other shards and nodes
  in the system.
- Pieces from a shard are uniquely identified through an sequentially increasing `piece_index`.
- Output: Segment with `NUM_PIECES` pieces and a `segment_root`.

#### 7. Segment proof Submission to Beacon Chain

- Input: New segments from a shard.
- As soon as a new block is produced that includes the creation of a new segment, the segment proof
  is submitted as a transaction in the beacon chain to commit the segment into the global history of
  the system.
- To submit this information to the beacon chain a `SegmentInfo` transaction is created that
  includes the `segment_root`, the `segment_index`, and the `shard_id` of the shard where the
  segment was created.
- Any node in the shard (from farmers to full-nodes) are allowed to create the `SegmentInfo`

>   transaction that submits the segments into the parent's shard history.

Reviewer: I'm not sure it shouldn't go through the parent shard first. Also it has the same problem that blocks have: it isn't protected from spamming in any way.

```rust

> type SegmentInfo = {
>   segment_root: Vec<u8>, // 32B
>   segment_index: u64, // 8B
>   shard_id: ShardId, // 4B
> }

Reviewer: ```suggestion
struct SegmentInfo {
  shard_id: ShardId, // 4B
  segment_index: u64,
  segment_root: [u8; 32],
}
```

Also segment index is currently a global thing, so it should probably be called differently or I'm not sure what you meant there, but child shard doesn't know the index until it is included in the beacon chain.

Author: I was assuming that segments will also keep a local index to determine their order within the shard, even if once they are committed they are assigned a global index.

Reviewer: Yes, that might be useful for syncing a specific shard indeed, but I'm not yet sure it is actually required for pushing up the stack

```

- Output: `SegmentInfo` transaction in the beacon chain mempool.

#### 8. Segment proof in beacon chain and Super Segment Construction

- Input: Transaction with `SegmentInfo` in the beacon chain.
- The execution of the `SegmentInfo` transaction when included in a block of the beacon chain

>   triggers the proof of the shard's segment in the global history of the beacon chain.

Reviewer: "triggers" feels off here

Author: Rephrased it a bit (hopefully it is clearer now).

- To verify the `SegmentInfo` and execute the transaction, nodes verify the following:

>   - Ensure `segment_index` is the subsequent index after the last one committed for the shard.

Reviewer: This is not really sufficient. I'd say all this should be happening as part of block header inclusion from the child shard, alongside with a proof that the segment was a part of the block (or place it in the header of the block, not sure).

- If the verification of the `SegmentInfo` is successful, the child shard segment is added to the
  beacon chain shard's history by adding the shard's `segment_root` to the `segment_roots[]` that

>   indexes the all proofs from the global history (including beacon chain and all child segments).

Reviewer: This doesn't make a lot of sense, the whole point of super segment is to avoid storing roots of all shards and instead compress it into a single super segment root.

- Apart from adding the segment proof in the global history, a new `SuperSegment` is created
  aggregating all the segment proofs included in a block of the beacon chain. The goal for this
  super segment proof is to simplify the verification of segments by light clients.
- The `super_segment` is created by aggregating the segment proofs of the child segments and
  creating a new proof that is included in the parent shard's history committing (and wrapping) all
  the history of shards in the lower layers of the architecture. This `super_segment` proof is
  directly a Merkle proof of all of the shard segments and child super segment proofs from that
  shard: `C_super_segment = Merkle_Proof(S1, ..., Sn, SS1, ..., SSn)`.
- Along with `C_super_segment` the corresponding `ws1` segments proof are created.

> - When a new super segment is created instead of broadcasting the network the full super segment,
>   only the `SuperSegmentHeader` is broadcasted, allowing nodes to proactively determine if they

Reviewer: ```suggestion
- When a new super segment is created instead of broadcasting to the network the full super segment,
  only the `SuperSegmentHeader` is broadcast, allowing nodes to proactively determine if they
```

I'm not sure how can nodes proactively determine just from a Merkle Root what commitments they have and what commitments they don't. In fact this `SuperSegmentHeadder` is never broadcast, it is composed locally by the block producer using segments they see (if they see any at all) in the transaction pool.

  require the full super segment data.
- The `SuperSegmentHeader` includes the following fields:
  - `history_delta`: Number of segment proofs included in the `super_segment`. This will be used to

>     determine the number of segments committed in the beacon chain at a specific `block_height`.

Reviewer: Delta can always be computed by comparing two segment headers, the only thing you need for knowing the size of the history (which is crucial for consensus) is the latest segment index that exists, which beacon chain by definition knows because it assigns them to shard segments.

  - `super_segment_root` proof of the super segment that implicitly verifies into the global history
    all the included segments.
  - `block_height` in the beacon chain where the super segment was created. This is used to identify
    the block in the child shard that contains the segments included in the super segment and pull
    all the super segment data (if needed).
- Output: Add segment proofs from children to `segment_roots[]`, and a new `SuperSegment` per block
  with segment proofs.

```rust
type SuperSegmentHeader = {
	super_segment_root: Vec<u8>,
  block_height: u64,
	history_delta: u64,
};
```

#### 9. Recursive Local Block and Segment Creation and Submission

- Input: Transactions with `BlockInfo`, new local segments, and `SegmentInfo`.
- The protocol is recursive, which means that if we have a hierarchical architecture with more than
  one level below the beacon chain, all shards independently of the level they belong to will

>   perform the same operations described in the sections from above: i.e. submit new

Reviewer: This is not exactly accurate. Super segments according to our latest discussion are only created on the beacon chain, so at least that is different.

  `RECORDED_HISTORY_SEGMENT_SIZE`.
- When added to the buffer, blocks are turned into `SegmentItem`s.
- Each segment will have the parent `segment_header` included as the first item. Each

>   `segment_header` includes the hash of the previous segment header and the segment proof of the

Author: Good catch, this was a typo. Thanks!
